Fix confusing code regaring handling of environment variables in compilerservice

arnaudsjs opened the pull request:
# Description

According to the docstring of `data.Compile`, the `used_environment_variables` field should be set to None until the compile starts. This was not in-line with the implementation. This in its turn made the interpretation of the `CompilerService._compile_merge_key()` method rather difficult. This PR fixes this issue by making the implementation in-line with the docstring.

# Self Check:

- [ ] ~~Attached issue to pull request~~
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] ~~End user documentation is included or an issue is created for end-user documentation~~
- [ ] ~~If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

sanderr reviewed (commented):
Looks like there's a type violation at line 431: it assumes that the value will never be `None`. Could you address that as well?

I'm still looking over the context for the rest of the review.

sanderr reviewed (approved):
While reviewing I came across [one concern](https://inmanta.slack.com/archives/CKRF0C8R3/p1712154130560919) but it doesn't directly impact the functionality of this PR or what we currently use this feature for.

arnaudsjs commented:
> While reviewing I came across [one concern](https://inmanta.slack.com/archives/CKRF0C8R3/p1712154130560919) but it doesn't directly impact the functionality of this PR or what we currently use this feature for.

I converted the assertion into an exception and added a test case for this scenario.

sanderr commented on src/inmanta/server/services/compilerservice.py:
```diff
@@ -650,7 +650,11 @@ async def request_recompile(
         shared_keys = mergeable_env_vars.keys() & env_vars.keys()
-        assert not shared_keys, f"An env var can not be both mergeable and normal: {shared_keys}"
+        if shared_keys:
+            raise Exception(
```
```suggestion
            raise ValueError(
```

I think this would be slightly more to the point, but I'm fine with either.

sanderr commented on tests/server/test_compilerservice.py:
```diff
@@ -1847,3 +1847,26 @@ async def wait_for_compile() -> bool:
+            env_vars={"var": "val", "test": "123"},
+            mergeable_env_vars={"var": "otherval", "somekey": "someval"},
+        )
+    assert (
+        "Invalid compile request: The same environment variable cannot be present in the "
+        "env_vars and mergeable_env_vars dictionary simultaneously: {'var'}." in str(exc.value)
+    )
```
FYI, `pytest.raises` has a `matches` keyword to match on the error message.

inmantaci commented:
Processing this pull request

inmantaci commented:
Pull request rejected by merge tool. The tests for this branch did not succeed.

inmantaci commented:
Processing this pull request

inmantaci commented:
Pull request rejected by merge tool. The tests for this branch did not succeed.

inmantaci commented:
Processing this pull request

inmantaci commented:
Merged into branches master in f2f66fcf21a23f8a0147e53ab8d37f19f41bcc4e

inmantaci commented:
Processing #7494.
